Add support for role/user based access

iamdharmesh opened the pull request:
### Description of the Change
This PR adds support for role- and user-based access control to the ClassifiAI feature. It also includes functionality for user-based opt-out, allowing users to choose not to use the specific feature if they prefer not to or if they dislike it.

**Feature settings:**
<img width="877" alt="image" src="https://github.com/10up/classifai/assets/10613171/de6b810b-b070-4751-86e6-41738cf9022a">

**User Profile:**
<img width="1020" alt="image" src="https://github.com/10up/classifai/assets/10613171/ba00f8fb-0c34-4d87-8794-8e724a14a44e">

**Edit User:**
<img width="1020" alt="image" src="https://github.com/10up/classifai/assets/10613171/fce99eff-5164-4aab-9bb3-0e3344dd4c74">

**In-Context disable feature link to user profile**
|Title generation|Resize content|
|---|---|
|![Screenshot 2023-10-30 at 5 17 15 PM](https://github.com/10up/classifai/assets/10613171/31daa294-61ac-45e4-b635-d579acbfaf60)|![Screenshot 2023-10-30 at 5 18 06 PM](https://github.com/10up/classifai/assets/10613171/aa9b63b7-c4a2-4419-98df-0ac9ab04dad3)|

_**Note:** E2E tests are not added for this new feature._

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #588

### How to test the Change
1. Enable role-based access to the feature.
1. Select roles in the allowed roles to provide access to that feature.
1. Verify that only the selected roles have access to that feature.
1. Enable user-based access to the feature.
1. Select the user in the allowed users to provide access to that feature.
1. Verify that the selected user has access to that feature.
1. Enable user-based opt-out for the feature.
1. Go to the user profile and verify that you see the checkbox for that feature to opt out of using it.
1. Check the opt-out checkbox and save.
1. Verify that access to that feature has been disabled.
1. Uncheck the opt-out checkbox to opt back in to the feature.
1. Verify that the user now has access to that feature.
1. Follow these steps for each ClassifAI feature.


### Changelog Entry
> Added - Role and user-based access control for features


### Credits
Props @jeffpaul @dkotter @iamdharmesh @Sidsector9 


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

iamdharmesh committed:
Add role-based access setting field in provider class.

iamdharmesh committed:
Made add_access_setting extendible for sub-feature as well.

iamdharmesh committed:
Use common access settings in ChatGPT.

iamdharmesh committed:
Add role-based access to classify content.

iamdharmesh committed:
Add role-based access to classify content (OpenAI)

iamdharmesh committed:
Added roll-based access to openAI whisper.

iamdharmesh committed:
Fix text to speech bulk actions and add role-based access to it.

iamdharmesh committed:
Add roll-based access to AI Vision

iamdharmesh committed:
Add additional permission check at rest endpoint for image processing.

iamdharmesh committed:
Added role-based access to image generation.

iamdharmesh committed:
Add feature names label in description text.

iamdharmesh committed:
Added role-based access to Recommended Content block.

iamdharmesh committed:
minor code improvements.

iamdharmesh committed:
Fixed is_feature_enabled call.

iamdharmesh committed:
standardized feature names with backward compatibility

iamdharmesh committed:
Added user-based access and user opt-out functionality.

iamdharmesh committed:
Moved access control to separate class.

iamdharmesh committed:
Changed since version number.

iamdharmesh committed:
Fix unit tests.

iamdharmesh committed:
Fix E2E tests.

iamdharmesh commented on includes/Classifai/Helpers.php:
```diff
@@ -772,3 +772,26 @@ function get_all_post_statuses() {
+		$feature . '_role_based_access'  => 1,
+		$feature . '_roles'              => array_keys( $editable_roles ),
+		$feature . '_user_based_access'  => 'no',
+		$feature . '_user_based_opt_out' => 'no',
```
@jeffpaul @dkotter Should user-based opt-out be enabled by default?

dkotter replied:
I don't think so, I'd leave it off by default

jeffpaul replied:
Agreed

dkotter commented on includes/Classifai/Providers/Watson/NLU.php:
```diff
@@ -88,63 +93,81 @@ public function __construct( $service ) {
 	}
 
+	/**
+	 * Default settings for ChatGPT
```
Think this needs updated from ChatGPT

iamdharmesh replied:
Thanks for catching this.

dkotter commented on includes/Classifai/Providers/Watson/NLU.php:
```diff
@@ -935,4 +962,36 @@ public function is_configured() {
+		/**
+		 * Filter to override permission to a IBM Watson classify content feature.
+		 *
+		 * @since 2.3.0
```
```suggestion
		 * @since 2.4.0
```

dkotter commented on includes/Classifai/Providers/Provider.php:
```diff
@@ -397,6 +430,44 @@ public function render_auto_caption_fields( $args ) {
+	 */
+	public function render_allowed_users( array $args = array() ) {
+		$setting_index = $this->get_settings();
+		$value         = ( isset( $setting_index[ $args['label_for'] ] ) ) ? $setting_index[ $args['label_for'] ] : array();
```
This can be simplified I think:

```suggestion
		$value         = $setting_index[ $args['label_for'] ] ?? array();
```

dkotter commented on includes/Classifai/Admin/UserProfile.php:
```diff
@@ -0,0 +1,183 @@
+	/**
+	 * Add ClassifAI features opt-out checkboxes to user profile and edit user.
+	 *
+	 * @param WP_User $user User object.
```
```suggestion
	 * @param \WP_User $user User object.
```

dkotter commented on includes/Classifai/Admin/UserProfile.php:
```diff
@@ -0,0 +1,183 @@
+	 * @param WP_User $user User object.
+	 * @return void
+	 */
+	public function user_settings( $user ) {
```
```suggestion
	public function user_settings( \WP_User $user ) {
```

dkotter commented on includes/Classifai/Providers/AccessControl.php:
```diff
@@ -0,0 +1,403 @@
+<?php
+/**
+ *  Class for handling the access control for ClassifAI features.
```
```suggestion
 * Class for handling the access control for ClassifAI features.
```

dkotter commented on includes/Classifai/Providers/AccessControl.php:
```diff
@@ -0,0 +1,403 @@
+		$this->user_based_opt_out_key = $feature . '_user_based_opt_out';
+		$this->users_key              = $feature . '_users';
+	}
+
```
Some extra spaces here that could be removed

dkotter commented on includes/Classifai/Providers/AccessControl.php:
```diff
@@ -0,0 +1,403 @@
+	 * @param Provider $provider The provider class instance.
+	 * @param string   $feature  The feature name.
+	 */
+	public function __construct( $provider, $feature ) {
```
```suggestion
	public function __construct( Provider $provider, string $feature ) {
```

dkotter commented on includes/Classifai/Providers/AccessControl.php:
```diff
@@ -0,0 +1,403 @@
+	 * @param string $section Settings section.
+	 * @return void
+	 */
+	public function add_settings( $section = '' ) {
```
```suggestion
	public function add_settings( string $section = '' ) {
```

dkotter commented on includes/Classifai/Providers/AccessControl.php:
```diff
@@ -0,0 +1,403 @@
+	 *
+	 * @return array The sanitized settings to be saved.
+	 */
+	public function sanitize_settings( $settings ) {
```
```suggestion
	public function sanitize_settings( array $settings ) {
```

dkotter commented on includes/Classifai/Helpers.php:
```diff
@@ -772,3 +772,26 @@ function get_all_post_statuses() {
+ * @param string $feature Feature key.
+ * @return array
+ */
+function get_feature_default_settings( $feature ) {
```
```suggestion
function get_feature_default_settings( string $feature ) {
```

dkotter commented on includes/Classifai/Providers/Provider.php:
```diff
@@ -465,4 +536,125 @@ public function is_configured() {
+	 * @param string $section Settings section.
+	 * @return void
+	 */
+	protected function add_access_settings( $feature, $section = '' ) {
```
```suggestion
	protected function add_access_settings( string $feature, string $section = '' ) {
```

dkotter commented on includes/Classifai/Providers/Provider.php:
```diff
@@ -465,4 +536,125 @@ public function is_configured() {
+	 *
+	 * @return array The sanitized settings to be saved.
+	 */
+	protected function sanitize_access_settings( $settings, $feature ) {
```
```suggestion
	protected function sanitize_access_settings( array $settings, string $feature ) {
```

dkotter commented:
Overall this looks really good, nice work here @iamdharmesh!

I've left a handful of minor code review comments. In addition to that, I have some thoughts on the UI and in particular some text we're showing.

In the settings for each feature, we have the following new text strings:

- Enables ability to select which role can access FEATURE
- Choose which roles are allowed to FEATURE
- Enables ability to select which users can access FEATURE
- Enables ability for users to opt-out from their user profile page

And then on the user profile page, we show:

- Opt out of using the FEATURE feature

For some features and settings, this text makes sense (like `Choose which roles are allowed to Classify Content`). But in most, the feature name doesn't pair well with this text (like `Enables ability to select which users can access Generate images`).

Open to ideas on any of these strings (especially from @jeffpaul) but I'd recommend the following:

- Since each of these strings is already within the settings section for a particular feature, I don't think we need to use the feature name itself. I'd suggest we standardize on:
     - Enables ability to select which roles can access this feature
     - Choose which roles are allowed to access this feature
     - Enables ability to select which users can access this feature

If we want to keep the feature name in there, I think we should change it to something like:

- Enables ability to select which role can access the FEATURE feature
- Choose which roles are allowed to access the FEATURE feature
- Enables ability to select which users can access the FEATURE feature

Again, I'd recommend the former over the latter but open to thoughts.

I think what we show on the profile page is fine, since it already has the `the FEATURE feature` part which makes more sense here in the context of a profile and not a settings page for the feature.

I would recommend though that we standardize on capitalization. Right now most features have the first word capitalized (like Generate captions) but some have both (like Classify Content). I suggest when using the feature name as a description, we should lower case all words. In addition, some of these descriptions have a trailing period and some don't. I think standardizing on no trailing period makes sense.

Another thing I'm seeing is that some of these settings descriptions don't have proper spacing (they're too close to the checkbox) while others are fine. See Watson NLU as an example:

![Screenshot 2023-10-27 at 12 25 50 PM](https://github.com/10up/classifai/assets/916738/a5e03b57-28ac-46e7-a014-240a190ebf4e)

Looks like there's some CSS scoped to `.classifai-content .classifai-wrap .classifai-input-description` that doesn't get applied everywhere. We should standardize on that to ensure we have proper spacing.

One area we already talked about but just mentioning again is this does make some of these settings pages pretty long (like Azure Image Processing). I think this is fine for now and will be helped once we complete the refactor to have stand-alone features in our settings (and even further once we do a refactor on these settings screens themselves) but just mentioning that there may be complaints on how unwieldy these pages are getting.

And I think final thing is looks like we are standardizing on `Image captions` being the feature name for the alt text generation. While I know we added that as a feature (where the alt text can be saved as a caption, description or alt text) I'm not sure this makes sense as the main feature name (also something for @jeffpaul to chime in on).

I think something like `Image descriptive text` instead of `Image captions` may make more sense, as well as `Generate descriptive text` instead of `Generate captions`

dkotter commented on src/js/admin.js:
```diff
@@ -78,3 +81,83 @@ document.addEventListener( 'DOMContentLoaded', function () {
+
+// Search for users.
+( () => {
+	jQuery( '.classifai-search-users' ).select2( {
```
Wondering if you looked at any Gutenberg components instead of using `select2` here? Thinking there may be a more native way to handle this. I know Gutenberg has an `Autocomplete` component, though not sure that allows you to populate from an API request. There's also a [`PostPicker` component](https://github.com/10up/ui-kit/tree/trunk/src/post-picker-block/blocks/post-picker) that 10up is working on that may provide some inspiration here.

Not saying this has to change but may be worth looking at alternatives that use built-in components

jeffpaul replied:
I agree on seeing what else is available as the select2 felt a bit off in terms of expected UI for that user selection.  @fabiankaegy any tips on your end for what could be used?

jeffpaul commented:
@dkotter @iamdharmesh I pretty much agree with Darin's recommendations in his https://github.com/10up/classifai/pull/606#issuecomment-1783343005

iamdharmesh committed:
Apply suggestions from code review

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>

iamdharmesh committed:
Update feature name from "Image captions" to "Image descriptive text" and "Generate captions" to "Generate descriptive text"

iamdharmesh committed:
Fixed settings description style consistency issue.

iamdharmesh committed:
Updated text strings of access control settings.

iamdharmesh committed:
Fix opt-out form submit in user profile.

iamdharmesh committed:
Added in-context disable ClassfAI feature links.

iamdharmesh committed:
Add disable feature link in classic editor.

iamdharmesh commented:
Thank you for the detailed review @dkotter.

> - Since each of these strings is already within the settings section for a particular feature, I don't think we need to use the feature name itself. I'd suggest we standardize on:
>   - Enables ability to select which roles can access this feature
>   - Choose which roles are allowed to access this feature
>   - Enables ability to select which users can access this feature

Updated description strings to follow this. Thanks for adding a note for this.

> I would recommend though that we standardize on capitalization. Right now most features have the first word capitalized (like Generate captions) but some have both (like Classify Content). I suggest when using the feature name as a description, we should lower case all words. In addition,

Done. 

> some of these descriptions have a trailing period and some don't. I think standardizing on no trailing period makes sense.

Standardized this on having a trailing period. existing field descriptions follow the period at the end. So, followed it. please let me know if you think we should update all to "no trailing period".

> Another thing I'm seeing is that some of these settings descriptions don't have proper spacing (they're too close to the checkbox) while others are fine. See Watson NLU as an example:
>
> ![Screenshot 2023-10-27 at 12 25 50 PM](https://github.com/10up/classifai/assets/916738/a5e03b57-28ac-46e7-a014-240a190ebf4e)
> Looks like there's some CSS scoped to `.classifai-content .classifai-wrap .classifai-input-description` that doesn't get applied everywhere. We should standardize on that to ensure we have proper spacing.

This was due to Watson NLU using an extended `render_input` function. Fixed. Thanks for catching this.

> One area we already talked about but just mentioning again is this does make some of these settings pages pretty long (like Azure Image Processing). I think this is fine for now and will be helped once we complete the refactor to have stand-alone features in our settings (and even further once we do a refactor on these settings screens themselves) but just mentioning that there may be complaints on how unwieldy these pages are getting.

We may explore quick win solutions here until we get a full refactor on these settings screens. Maybe something like vertical tabs for each section and only active section will be shown. This will help make the settings page smaller and organized. Obviously, this is not a perfect solution but something we can explore until we work on refactoring.


> And I think final thing is looks like we are standardizing on `Image captions` being the feature name for the alt text generation. While I know we added that as a feature (where the alt text can be saved as a caption, description or alt text) I'm not sure this makes sense as the main feature name (also something for @jeffpaul to chime in on).
>
> I think something like `Image descriptive text` instead of `Image captions` may make more sense, as well as `Generate descriptive text` instead of `Generate captions`

Updated accordingly. Thank you.

iamdharmesh committed:
Fixed codeQL warning.

dkotter commented on src/js/components/disable-feature-button.js:
```diff
@@ -0,0 +1,34 @@
+
+	return (
+		<Button
+			href={ ClassifAI?.profile_url }
```
Would be great to link down to the ClassifAI section, which I think should be as simple as adding `#classifai-profile-features-section` to the end of this URL (either here or where we localize this data)

iamdharmesh replied:
done.

iamdharmesh committed:
Removed select2.

iamdharmesh committed:
Added UserSelector component.

iamdharmesh committed:
Integrate user selector in classifai settings.

iamdharmesh committed:
Some design tweaks.

iamdharmesh committed:
Fix suggestion background color.

jeffpaul commented:
@iamdharmesh besides resolving merge conflicts, is there any additional work here before this can re-route for Darin's code review?

iamdharmesh committed:
Fix user selection glitch.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/588

iamdharmesh committed:
remove debug code.

iamdharmesh committed:
Fix e2e in wp-minimum env.

iamdharmesh commented:
@jeffpaul @dkotter This is ready for review now.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/588

berkod commented on includes/Classifai/Admin/BulkActions.php:
```diff
@@ -177,17 +179,16 @@ public function register_bulk_actions( $bulk_actions ) {
-			'no' !== $computer_vision_settings['enable_image_tagging'] ||
-			! empty( $this->computer_vision->get_alt_text_settings() )
+			$this->computer_vision->is_feature_enabled( 'image_tagging' ) ||
+			$this->computer_vision->is_feature_enabled( 'image_captions' )
```
is it clear in instructions that this tagging and captions go together?

iamdharmesh committed:
Oraganized language processing e2e tests.

iamdharmesh committed:
Added E2E tests around access control for language processing.

iamdharmesh committed:
Organized image processing tests and added access control tests for image generation.

iamdharmesh committed:
Add access control tests for AI Vision features.

iamdharmesh committed:
Added test isolation and before scripts.

iamdharmesh committed:
Add command for enable classic editor.

iamdharmesh committed:
Fix title and excerpt generation tests.

iamdharmesh committed:
Fix lodash dependency error.

iamdharmesh committed:
Fix tests on wp minimum env.

iamdharmesh committed:
Lint fixes in e2e tests.

iamdharmesh commented:
@dkotter I have added E2E tests, could you please help with re-reviewing this? Thanks

iamdharmesh committed:
Add enable feature checkbox in classify content (IBM watson).

iamdharmesh committed:
Update Onboarding feature to classify content.

iamdharmesh committed:
Add enable/disable feature support in Text to speech.

iamdharmesh committed:
Added is_enabled method for feature.

iamdharmesh committed:
Add enable/disable feature to recommended content.

iamdharmesh committed:
Add E2E tests for enable/disable feature.

iamdharmesh committed:
Fix conflict warning on save classify content.

jeffpaul commented:
@iamdharmesh one item that came up in UAT today was the user opt-out links from within the editor/etc. is missing on the Generate Images page linked off the Media Library admin menu.  Please ensure we add the opt-out link to the user profile there (if the opt-out setting is allowed) and double-check that we're adding that link across all the ClassifAI features... thanks!

iamdharmesh committed:
Add feature disable link to profile in media model.

iamdharmesh committed:
render disable feature link markup using helper function.

iamdharmesh committed:
Add disable feature link to generate excerpt feature.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/588

iamdharmesh commented:
@jeffpaul I have added a disable feature link to generate images popup and Generate excerpt metabox. Also, please find the status of each feature for the disable link. I have skipped most of the automated features and added a potential UI option to add a link if needed. Could you please help to check and please let me know if you think we should add disable links for the given features? We may handle it in follow-up PR to unblock this PR to be merged.

### Language Processing:
- **IBM Watson (Natural Language Understanding):** 
Skipped as we have confirmation popup development in progress. (disable link should be added there)

- **OpenAI Chatgpt**
    - **Excerpt Generation:** ✅ 
    - **Title Generation:** ✅ 
    - **Resize content:** ✅ 
    
- **OpenAI Embeddings:**
Skipped due to automated behavior

- **OpenAI Whisper:** 
Skipped due to automated behavior.
But we can add a link to the media model if needed (though it will make the UI a bit messy), what do you think?
![Screenshot 2023-11-17 at 12 15 51 PM](https://github.com/10up/classifai/assets/10613171/d7c81521-319d-4bf2-a615-4f5a7b1628b9)

- **Microsoft Azure (Text to Speech):** 
Skipped due to automated behavior.
We can add disable link in metabox and ClassifAI slotfill. what do you think?
![image](https://github.com/10up/classifai/assets/10613171/41a16f76-2676-41bc-a0ed-de5f9da5416d)


### Image Processing:
- **AI Vision( Microsoft Azure):**
It skipped due to automated behavior, but we can add a link to the media model if needed.
![image](https://github.com/10up/classifai/assets/10613171/e6451b32-c475-4e11-a9cc-8f218a3d14c6)


- **OpenAI DALL·E (Image generation)**:  ✅ 

cc: @dkotter

jeffpaul commented:
I think adding the disable link to all the places you mention are worthwhile, but defer if @dkotter thinks any aren't necessary.  Also, if you could please push a commit to #609 for the post classification disabling that would be helpful... thanks!

dkotter commented:
Yeah, if we have places we can easily add this messaging, I think it's worth doing

dkotter committed:
Merge pull request #614 from 10up/enhancement/589

Add support controlling feature availability

dkotter committed:
Add an older version of mochawesome-json-to-md as the latest version is broken
